src/java/org/apache/cassandra/index/sai/plan/Plan.java

```diff
@@ -1208,14 +1252,44 @@ protected KeyRangeIterator execute(Executor executor)
             }
         }
 
+        @Override
+        protected boolean usesIncludedIndex()
+        {
+            return subplansSupplier.get().stream().anyMatch(KeysIteration::usesIncludedIndex);
+        }
+
         /**
          * Limits the number of intersected subplans
          */
         public Plan stripSubplans(int clauseLimit)
         {
             if (subplansSupplier.orig.size() <= clauseLimit)
                 return this;
-            List<Plan.KeysIteration> newSubplans = new ArrayList<>(subplansSupplier.orig.subList(0, clauseLimit));
+
+            if (factory.hints.included.isEmpty())
+                return withNewSubplans(new ArrayList<>(subplansSupplier.orig.subList(0, clauseLimit)));
+
+            List<Plan.KeysIteration> newSubplans = new ArrayList<>(clauseLimit);
+            List<Plan.KeysIteration> optionalSubplans = new ArrayList<>(clauseLimit);
+            for (KeysIteration keysIteration : subplansSupplier.orig)
+            {
+                if (keysIteration.usesIncludedIndex())
+                    newSubplans.add(keysIteration);
+                else
+                    optionalSubplans.add(keysIteration);
+            }
+            for (KeysIteration keysIteration : optionalSubplans)
+            {
+                if (newSubplans.size() >= clauseLimit)
+                    break;
+                optionalSubplans.add(keysIteration);
```

Comment:
This doesn't look right: adding the keys iteration again to the optional subplans. Shouldn't be?:
```suggestion
                newSubplans.add(keysIteration);
```
Can you add a test to catch this bug?

Reply:
It's indeed a bug! I have fixed it and modified `PlanWithIndexHintsTest#testIntersectionClauseLimit` to exercise this part of the code.